go/proxyd/backend.go

```diff
@@ -199,51 +306,219 @@ func (b *Backend) doForward(rpcReq *RPCReq) ([]byte, error) {
 
 type BackendGroup struct {
 	Name     string
-	backends []*Backend
-	i        int64
+	Backends []*Backend
 }
 
 func (b *BackendGroup) Forward(rpcReq *RPCReq) (*RPCRes, error) {
-	var outRes *RPCRes
-	for _, back := range b.backends {
+	rpcRequestsTotal.Inc()
+
+	for _, back := range b.Backends {
 		res, err := back.Forward(rpcReq)
-		if err == ErrBackendOffline {
+		if errors.Is(err, ErrMethodNotWhitelisted) {
+			return nil, err
+		}
+		if errors.Is(err, ErrBackendOffline) {
 			log.Debug("skipping offline backend", "name", back.Name)
 			continue
 		}
+		if errors.Is(err, ErrBackendOverCapacity) {
+			log.Debug("skipping over-capacity backend", "name", back.Name)
+			continue
+		}
 		if err != nil {
 			log.Error("error forwarding request to backend", "err", err, "name", b.Name)
 			continue
 		}
-		outRes = res
-		break
+		return res, nil
+	}
+
+	return nil, ErrNoBackends
+}
+
+func (b *BackendGroup) ProxyWS(clientConn *websocket.Conn) (*WSProxier, error) {
+	for _, back := range b.Backends {
+		proxier, err := back.ProxyWS(clientConn)
+		if errors.Is(err, ErrBackendOffline) {
```

Comment:
Does this need to handle the `ErrMethodNotWhitelisted` case as well?

Reply:
It does - see the implementation of `WSProxier`. We can't capture that error here because the WS connection happens before we start receiving messages.